search/includes/classes/class-cache.php

```diff
@@ -0,0 +1,71 @@
+<?php
+namespace Automattic\VIP\Search;
+
+class Cache {
+	public function __construct() {
+		add_action( 'pre_get_posts', array( $this, 'disable_apc_for_ep_enabled_requests' ), 0 );
+	}
+
+	/**
+	 * Advanced Post Cache and ElasticPress do not work well together.
+	 * APC caches post IDs populates the $wp_query->posts by running `get_post` on each cached post id during `posts_request`,
+	 * ElasticPress runs on `posts_pre_query` which fires after `posts_request`. Here we disable APC if this query is offloaded to EP.
+	 * 
+	 *
+	 * On the other hand, if a non-ElasticPress query is run, and we disabled
+	 * Advanced Post Cache earlier, we enable it again, to make use of its caching
+	 * features.
+	 *
+	 *
+	 * @param WP_Query $query The query to examine.
+	 */
+	public function disable_apc_for_ep_enabled_requests( &$query ) {
+		global $advanced_post_cache_object;
+
+		static $disabled_apc = false;
+
+		if ( ! is_a( $advanced_post_cache_object, 'Advanced_Post_Cache' ) ) {
+			return;
+		}
+
```

Comment:
Should we also check that `$query` is an instance of `WP_Query`, just to be safe?

Reply:
I thought about it, but ultimately figured that `\ElasticPress\Indexables::factory()->get( 'post' )->elasticpress_enabled( $query )` check wouldn't pass for any other type of query, so a separate check would be redundant? (That may be not true but just wishful thinking.)

Reply:
It looks like `Indexable::elasticpress_enabled()` doesn't do any checking like that, sadly. Best to be defensive in our code anyway, in case EP were to change the implementation on us.

Reply:
The check's would probably still return `false` and not _break_ at least.

Reply:
Fair point.